core/src/main/java/com/graphhopper/routing/weighting/custom/CustomWeighting.java

```diff
@@ -0,0 +1,152 @@
+ * speed = reduce_to_max_speed(estimated_average_speed * multiply_all(speed_factor_map))
+ * weight = (toSeconds(distance / speed) + distance * distance_term_constant) / priority;
+ * return weight
+ * </pre>
+ * Please note that the max_speed map is capped to the maximum allowed speed. Also values in the speed_factor and
+ * priority maps are normalized via the maximum value to 1 if one of the values is bigger than 1
+ * to avoid problems with the landmark algorithm, i.e. the edge weight is always increased and the heuristic always
+ * underestimates the weight.
+ */
+public final class CustomWeighting extends AbstractWeighting {
+
+    /**
+     * @return the weighting name from the modelName
+     */
+    public static String key(String modelName) {
+        return "custom_" + modelName;
+    }
+
+    /**
+     * @return the modelName from the weigting
+     */
+    public static String modelName(String weighting) {
+        return weighting.substring(key("").length());
+    }
+
+    private final BooleanEncodedValue baseVehicleProfileAccessEnc;
+    private final String baseVehicleProfile;
+    private final double maxSpeed;
+    private final double minDistanceTerm;
+    private final SpeedCustomConfig speedConfig;
+    private final PriorityCustomConfig priorityConfig;
+
+    public CustomWeighting(String name, FlagEncoder baseFlagEncoder, EncodedValueLookup lookup,
+                           EncodedValueFactory factory, TurnCostProvider turnCostProvider, CustomModel customModel) {
+        super(key(name), baseFlagEncoder, turnCostProvider);
+        baseVehicleProfileAccessEnc = baseFlagEncoder.getAccessEnc();
+        baseVehicleProfile = customModel.getBase();
+
+        speedConfig = new SpeedCustomConfig(baseFlagEncoder.getMaxSpeed(), customModel, lookup, factory);
+        maxSpeed = speedConfig.getMaxSpeed() / CustomModel.SPEED_CONV;
+
+        priorityConfig = new PriorityCustomConfig(customModel, lookup, factory);
+
+        minDistanceTerm = customModel.getDistanceTermConstant();
+        if (minDistanceTerm < 0)
+            throw new IllegalArgumentException("maximum distance_term cannot be negative " + minDistanceTerm);
+    }
+
+    /**
+     * This method sets the vehicle of the specified request. It uses the importCustomModels if customModel is null.
+     */
+    public static CustomModel prepareRequest(GHRequest request, CustomModel customModel, Map<String, CustomModel> importCustomModels) {
+        HintsMap hints = request.getHints();
+        if (hints.getWeighting().startsWith(CustomWeighting.key(""))) {
+            if (customModel == null) {
+                String modelName = modelName(hints.getWeighting());
+                if ((customModel = importCustomModels.get(modelName)) == null)
+                    throw new IllegalArgumentException("unknown custom model " + modelName + " (" + hints.getWeighting() + ")");
+            }
+            request.setVehicle(customModel.getBase());
```

Comment:
We use the base as vehicle, but what if this custom model uses another custom model as base? Or maybe this feature is not implemented yet? Anyway this will be done differently as soon as we have the profile names (which I think should go to master before this PR or what do you think?)

Reply:
Also we will (later) merge customModel with the importCustomModel here, right?

Reply:
Currently only FlagEncoders work as `base` and yes: the profile name PR should come before this here and yes: the merging part is not yet implemented - see the open tasks :)